File: backend/app/schemas.py
from __future__ import annotations
from typing import Literal
from pydantic import BaseModel, Field

VerifiedStatus = Literal["confirmed", "needs_review", "unfounded"]


class SourceDoc(BaseModel):
    id: str
    cycle_id: str
    title: str
    text: str


class Citation(BaseModel):
    quote: str
    source_id: str


class DiagnosisItem(BaseModel):
    id: str
    channel: str
    summary: str
    kind: Literal["strength", "weakness"]
    citations: list[Citation] = Field(default_factory=list)
    status: VerifiedStatus = "needs_review"


class OpportunityRiskItem(BaseModel):
    id: str
    kind: Literal["opportunity", "risk"]
    title: str
    rationale: str
    citations: list[Citation] = Field(default_factory=list)
    additionally_flagged: bool = False


class CriticalPoint(BaseModel):
    id: str
    title: str
    impact: str
    urgency: str
    decision_needed: str
    citations: list[Citation] = Field(default_factory=list)


class TimelineLink(BaseModel):
    item_title: str
    prior_cycle_id: str
    same_issue: bool
    rebuttal_passed: bool
    repeat_count: int = 1


class ActionItem(BaseModel):
    id: str
    title: str
    owner: str
    due: str
    priority: Literal["high", "mid", "low"]
    # knowledge-work-plugins의 marketing/performance-report·sales/pipeline-review
    # 스킬이 쓰는 Impact/Effort 매트릭스 패턴 — priority만으로는 "왜 급한가"와
    # "하기 쉬운가"가 섞여 보이지 않는다.
    impact: Literal["high", "mid", "low"] = "mid"
    effort: Literal["high", "mid", "low"] = "mid"
    source_item_ids: list[str] = Field(default_factory=list)


class ActionItemsReport(BaseModel):
    """Action Items — 즉시 확인(긴급) → 조치 필요(일반) → 최종 요약 순으로 구성."""

    immediate_check: list[ActionItem] = Field(default_factory=list)
    action_needed: list[ActionItem] = Field(default_factory=list)
    final_summary: str = ""


class CustomerStrategy(BaseModel):
    """고객(법인)별 대응 전략 — 특정 고객/세그먼트에 대한 현황과 대응 방향."""

    id: str
    customer: str
    situation: str
    strategy: str
    citations: list[Citation] = Field(default_factory=list)


class ProcessStep(BaseModel):
    """법인 대응 process의 한 단계. 절차적 산출이라 축자 인용을 요구하지 않는다."""

    order: int
    title: str
    description: str
    owner: str = ""


class MetricSnapshot(BaseModel):
    """성과 지표 스냅샷 — knowledge-work-plugins의 marketing/performance-report,
    moai-cowork의 marketing-performance-report 스킬이 쓰는 "측정 → 관리 → 개선"
    대시보드 패턴. current/prior/target은 원문에 있는 숫자만 담는다 — 시장
    평균치 같은 외부 벤치마크는 원문에 없으면 만들어 넣지 않는다."""

    metric: str
    current: str
    prior: str = ""
    change: str = ""
    target: str = ""
    status: Literal["on_track", "at_risk", "off_track"] = "at_risk"


class DiagnosisSummary(BaseModel):
    """현황진단을 임원 의사결정 형식으로 종합한 결과."""

    executive_summary: str = ""
    metrics: list[MetricSnapshot] = Field(default_factory=list)
    customer_strategies: list[CustomerStrategy] = Field(default_factory=list)
    corporate_response_process: list[ProcessStep] = Field(default_factory=list)


class StrategicAxis(BaseModel):
    """전략의 3축 — 이번 회차 진단을 관통하는 핵심 전략 방향 하나."""

    id: str
    title: str
    description: str
    citations: list[Citation] = Field(default_factory=list)


class IssueStrategyGuide(BaseModel):
    """사안별 전략 가이드 — 특정 이슈(상충·리스크·CP)에 대한 대응 방향."""

    id: str
    issue_title: str
    guide: str
    source_item_ids: list[str] = Field(default_factory=list)
    citations: list[Citation] = Field(default_factory=list)


class RecommendedTimelineStep(BaseModel):
    """권장 타임라인의 한 단계 — 앞으로 실행을 언제·무엇을·누가 할지 제안."""

    order: int
    when: str
    action: str
    owner: str = ""


class StrategyTimeline(BaseModel):
    """전략/타임라인 — 사안별 전략 가이드 + 전략의 3축 + 권장 타임라인."""

    issue_guides: list[IssueStrategyGuide] = Field(default_factory=list)
    strategic_axes: list[StrategicAxis] = Field(default_factory=list)
    recommended_timeline: list[RecommendedTimelineStep] = Field(default_factory=list)


class CycleReport(BaseModel):
    cycle_id: str
    diagnosis: list[DiagnosisItem] = Field(default_factory=list)
    opportunities_risks: list[OpportunityRiskItem] = Field(default_factory=list)
    critical_points: list[CriticalPoint] = Field(default_factory=list)
    diagnosis_summary: DiagnosisSummary = Field(default_factory=DiagnosisSummary)
    timeline: list[TimelineLink] = Field(default_factory=list)
    strategy_timeline: StrategyTimeline = Field(default_factory=StrategyTimeline)
    action_items: ActionItemsReport = Field(default_factory=ActionItemsReport)
    overview: str = ""
    overview_warnings: list[str] = Field(default_factory=list)
    coverage_note: str = ""

File: frontend/lib/schemas.ts
import { z } from "zod";

/**
 * backend/app/schemas.py와 1:1로 대응하는 zod 스키마. API 응답을 파싱 시점에
 * 검증해, 백엔드 스키마가 바뀌었는데 프론트가 조용히 깨지는 것(런타임에야 발견되는
 * undefined 접근)을 막는다 — 이 세션에서만 CycleReport 모양이 세 번 바뀌었다.
 */

export const SourceTypeSchema = z.enum([
  "email", "social", "crm", "analytics", "customer_feedback", "news", "upload", "manual",
]);

export const SOURCE_TYPE_LABEL: Record<z.infer<typeof SourceTypeSchema>, string> = {
  email: "이메일 캠페인",
  social: "소셜미디어",
  crm: "CRM/영업 파이프라인",
  analytics: "웹/앱 애널리틱스",
  customer_feedback: "고객 피드백/VoC",
  news: "뉴스/경쟁사 동향",
  upload: "업로드 문서",
  manual: "직접 입력",
};

export const CitationSchema = z.object({
  quote: z.string(),
  source_id: z.string(),
});

export const DiagnosisItemSchema = z.object({
  id: z.string(),
  channel: z.string(),
  summary: z.string(),
  kind: z.enum(["strength", "weakness"]),
  citations: z.array(CitationSchema),
  status: z.enum(["confirmed", "needs_review", "unfounded"]),
});

export const OpportunityRiskItemSchema = z.object({
  id: z.string(),
  kind: z.enum(["opportunity", "risk"]),
  title: z.string(),
  rationale: z.string(),
  citations: z.array(CitationSchema),
  additionally_flagged: z.boolean(),
});

export const CriticalPointSchema = z.object({
  id: z.string(),
  title: z.string(),
  impact: z.string(),
  urgency: z.string(),
  decision_needed: z.string(),
  citations: z.array(CitationSchema),
});

export const TimelineLinkSchema = z.object({
  item_title: z.string(),
  prior_cycle_id: z.string(),
  same_issue: z.boolean(),
  rebuttal_passed: z.boolean(),
  repeat_count: z.number(),
});

const LevelSchema = z.enum(["high", "mid", "low"]);

export const ActionItemSchema = z.object({
  id: z.string(),
  title: z.string(),
  owner: z.string(),
  due: z.string(),
  priority: LevelSchema,
  impact: LevelSchema,
  effort: LevelSchema,
  source_item_ids: z.array(z.string()),
});

export const ActionItemsReportSchema = z.object({
  immediate_check: z.array(ActionItemSchema),
  action_needed: z.array(ActionItemSchema),
  final_summary: z.string(),
});

export const CustomerStrategySchema = z.object({
  id: z.string(),
  customer: z.string(),
  situation: z.string(),
  strategy: z.string(),
  risk_level: z.enum(["stable", "at_risk", "critical"]),
  citations: z.array(CitationSchema),
});

export const ProcessStepSchema = z.object({
  order: z.number(),
  title: z.string(),
  description: z.string(),
  owner: z.string(),
});

export const MetricSnapshotSchema = z.object({
  metric: z.string(),
  current: z.string(),
  prior: z.string(),
  change: z.string(),
  target: z.string(),
  status: z.enum(["on_track", "at_risk", "off_track"]),
});

export const DiagnosisSummarySchema = z.object({
  executive_summary: z.string(),
  metrics: z.array(MetricSnapshotSchema),
  customer_strategies: z.array(CustomerStrategySchema),
  corporate_response_process: z.array(ProcessStepSchema),
});

export const StrategicAxisSchema = z.object({
  id: z.string(),
  title: z.string(),
  description: z.string(),
  citations: z.array(CitationSchema),
});

export const IssueStrategyGuideSchema = z.object({
  id: z.string(),
  issue_title: z.string(),
  guide: z.string(),
  source_item_ids: z.array(z.string()),
  citations: z.array(CitationSchema),
});

export const RecommendedTimelineStepSchema = z.object({
  order: z.number(),
  when: z.string(),
  action: z.string(),
  owner: z.string(),
});

export const StrategyTimelineSchema = z.object({
  issue_guides: z.array(IssueStrategyGuideSchema),
  strategic_axes: z.array(StrategicAxisSchema),
  recommended_timeline: z.array(RecommendedTimelineStepSchema),
});

export const CycleReportSchema = z.object({
  cycle_id: z.string(),
  diagnosis: z.array(DiagnosisItemSchema),
  opportunities_risks: z.array(OpportunityRiskItemSchema),
  critical_points: z.array(CriticalPointSchema),
  diagnosis_summary: DiagnosisSummarySchema,
  timeline: z.array(TimelineLinkSchema),
  strategy_timeline: StrategyTimelineSchema,
  action_items: ActionItemsReportSchema,
  overview: z.string(),
  overview_warnings: z.array(z.string()),
  coverage_note: z.string(),
});

export const SourceDocSchema = z.object({
  id: z.string(),
  cycle_id: z.string(),
  title: z.string(),
  text: z.string(),
  source_type: SourceTypeSchema,
});
export const SourceDocListSchema = z.array(SourceDocSchema);

export const AddSourceResponseSchema = z.object({ id: z.string() });
export const CycleListSchema = z.array(z.string());

export type SourceType = z.infer<typeof SourceTypeSchema>;
export type SourceDoc = z.infer<typeof SourceDocSchema>;
export type Citation = z.infer<typeof CitationSchema>;
export type DiagnosisItem = z.infer<typeof DiagnosisItemSchema>;
export type OpportunityRiskItem = z.infer<typeof OpportunityRiskItemSchema>;
export type CriticalPoint = z.infer<typeof CriticalPointSchema>;
export type TimelineLink = z.infer<typeof TimelineLinkSchema>;
export type ActionItem = z.infer<typeof ActionItemSchema>;
export type ActionItemsReport = z.infer<typeof ActionItemsReportSchema>;
export type CustomerStrategy = z.infer<typeof CustomerStrategySchema>;
export type ProcessStep = z.infer<typeof ProcessStepSchema>;
export type MetricSnapshot = z.infer<typeof MetricSnapshotSchema>;
export type DiagnosisSummary = z.infer<typeof DiagnosisSummarySchema>;
export type StrategicAxis = z.infer<typeof StrategicAxisSchema>;
export type IssueStrategyGuide = z.infer<typeof IssueStrategyGuideSchema>;
export type RecommendedTimelineStep = z.infer<typeof RecommendedTimelineStepSchema>;
export type StrategyTimeline = z.infer<typeof StrategyTimelineSchema>;
export type CycleReport = z.infer<typeof CycleReportSchema>;
